### PR TITLE
Text localization examples

### PR DESCRIPTION
**Pull request by Maatteogekko**
Added Text l10n examples following @suragch structure.

Screenshot 2024-01-02 at 08 32 17
Screenshot 2024-01-02 at 08 32 29
Screenshot 2024-01-02 at 08 32 41
Screenshot 2024-01-02 at 08 32 59

### CODE REVIEW

**Comment by Maatteogekko** [comment]
@matthew-carroll let me know if creating a branch was a good move, or if pulling from a fork would have been more appropriate

**Comment by suragch** [comment]
As I've been looking at attributed strings to work on #8, I see that attributed strings can also be localized. I'm not sure if this should be included or not, but you can see the docs [here](https://developer.apple.com/documentation/foundation/attributedstring/3867590-init).

**Comment by Maatteogekko** [comment]
yes I will need to add some examples on that. They should be treated separately from Text though, since they can be used in many other Views and also outside of UI code.
I propose treating these cross-view APIs separately, as if they were separate Views. WDYT?

Or we can bundle them up with the first View that uses them, like in this case, and then refactor them later

**Comment by suragch** [comment]
> I propose treating these cross-view APIs separately, as if they were separate Views. WDYT?
> 
> Or we can bundle them up with the first View that uses them, like in this case, and then refactor them later

I'd vote for bundling them for now so that we get a better idea of what we are dealing with. But I see your point about them being separate. In my current PR I added attributed string examples but didn't try to exhaustively replicate the attributed string API.

**Comment by Maatteogekko** [comment]
Makes sense. Then I'll also add examples for AttributedString

**Comment by matthew-carroll** [comment]
@Maatteogekko thanks for the PR. Looking at your screenshots, I notice that there's no navigation bar. Are these examples accessible through the example app?

Also, looking at the content in the example, it's difficult to know what behavior you're demonstrating. Can you include a brief description at the top of the example page that explains what we're looking at in that example?

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift [hunk 1]
```diff
@@ -1,26 +1,49 @@
...
         ScrollView {
             VStack(spacing: 24) {
-                // Localized if possible
+                // Text's initializer accepts a LocalizedStringKey, which can be implicitly constructed from a string literal
+                // The value of the string is used for the development language localization
+                // Values for other languages are looked up in the default table using the string as key, if not found the value is used as is
                 Text("pencil")
                 
-                // Variables not localized by default
+                // The comment is shown in the localization table, useful for giving context to translators
```
Comment (matthew-carroll) [inline doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:13]: In this comment, you mentioned a "localization table" showing the "comment" mentioned in this `Text` view, i.e., the comment is placed in some kind of table at runtime.

However, down below, you reference the `.main` localization table, which suggests that the table exists before the code is compiled, not after, and not at runtime.

Can you clarify what you mean by "localization table"?

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift [hunk 2]
```diff
@@ -1,26 +1,49 @@
...
+                Text("another pencil", comment: "Translator comment")
+                
+                // This comment will be ignored, since the same key was declared earlier. This is valid even if no comment was specified on the previous key.
```
Comment (matthew-carroll) [inline doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:16]: What happens if a `Text` view has a paragraph of text? Is all of that a single key?

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift [hunk 3]
```diff
@@ -1,26 +1,49 @@
...
+                Text("another pencil", comment: "A different comment")
+                
+                // Specify a different table for this localization
+                Text("pencil", tableName: "LocalizableAlternative")
```
Comment (matthew-carroll) [inline doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:20]: Are there any naming requirements for the table name? Do they have to be "LocalizableXXXX"?

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift (hunk 1 above)
Comment (Maatteogekko) [reply doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:30]: I added a comment about table generation, let me know if it still isn't clear

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift (hunk 2 above)
Comment (Maatteogekko) [reply doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:33]: yep, added an example to illustrate this

File: doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift (hunk 3 above)
Comment (Maatteogekko) [reply doc_swift_ui_gallery/swift_ui_gallery/swift_ui_gallery/primitives/text/localization/TextLocalizationPage.swift:37]: LocalizableXXX is the naming convention for string catalogs. The only requirement here is that it is the name of an existing string catalog.
I added to the comment to make this clearer

**Comment by Maatteogekko** [comment]
Tomorrow I'll add the examples on AttributedStrings

**Comment by Maatteogekko** [comment]
> Looking at your screenshots, I notice that there's no navigation bar. Are these examples accessible through the example app?

The screens were taken from the preview, which has no ancestor View, so the navigation bar was not displayed. I updated them

**Comment by matthew-carroll** [comment]
@Maatteogekko how are things going with this PR?

**Comment by Maatteogekko** [comment]
I am having trouble with getting AttributedString to localize properly. Apart from that, if everything else is good, we can merge this and I'll add AttributedStrings in another PR

**Comment by matthew-carroll** [comment]
Can you describe what it is you're trying to do, and what trouble you're running into?

**Comment by Maatteogekko** [comment]
```swift
let attributed = AttributedString(localized: "pencil")
Text(attributed)
```

The localization value for "pencil" ("matita") is not displayed in the UI. I haven't been able to find a simple example that shows this working online. My guess is that AttributedString does not (yet?) work with string catalogs. I will try using string dicts, which are now legacy, and see if I get any luck with that.

**Comment by Maatteogekko** [comment]
Finally solved the mystery with AttributedString: the localization for AttributedStrings is only applied for the system locale, go figure.
I couldn't find any explanation for this in the official documentation. Shoutout to https://fatbobman.com/en/posts/attributedstring for providing the solution (and a lot of useful examples).

I also added a grammar agreement example, which is just an introduction to some of the automatic localization mechanisms offered by swiftui

**Comment by matthew-carroll** [comment]
@Maatteogekko would you mind posting updated screenshots?

**Comment by Maatteogekko** [comment]
@matthew-carroll done

**Review by matthew-carroll — APPROVED** [review]
LGTM - Can you now take what you learned in this process and fill out the section on Localization in the spec document? Please use Jonathan's write-up as an example of the level of detail.

WRT "string catalogs" and "localization tables" - if these mean the same thing then please choose one term and use it consistently. Otherwise, if they mean two different things, please clearly define the terms before using them in the writeup.

**Pull request merged**